### ISSUE TITLE
torch.det does not compute gradients correctly

### ISSUE THREAD

**Issue by finnoshea** [pytorch/pytorch#104675]
### 🐛 Describe the bug

Gradients are not properly computed when [torch.det](https://pytorch.org/docs/stable/generated/torch.det.html) is used.  I am using pytorch version 2.0.1+cu117.

```
import torch

print('---------torch.det----------')


x1 = torch.zeros(2,1, dtype=torch.float, requires_grad=True)
x2 = torch.ones(2,1, dtype=torch.float, requires_grad=True)
x = torch.hstack((x1,x2))
x.retain_grad()
y = torch.det(x)
y.backward()
print('input matrix:')
print(x)
print('gradients on input matrix:')
print(x.grad)


print('---------torch..eig---------')


w1 = torch.zeros(2,1, dtype=torch.float, requires_grad=True)
w2 = torch.ones(2,1, dtype=torch.float, requires_grad=True)
w = torch.hstack((w1,w2))
w.retain_grad()
z = torch.prod(torch.linalg.eigvals(w))
z.backward()
print('input matrix:')
print(w)
print('gradients on input matrix:')
print(w.grad)
```

The output:

> ---------torch.det----------
> input matrix:
> tensor([[0., 1.],
>         [0., 1.]], grad_fn=<CatBackward0>)
> gradients on input matrix:
> tensor([[0., -0.],
>         [-0., 0.]])
> ---------torch..eig---------
> input matrix:
> tensor([[0., 1.],
>         [0., 1.]], grad_fn=<CatBackward0>)
> gradients on input matrix:
> tensor([[ 1.,  0.],
>         [-1.,  0.]])

The second method, using `torch.linalg.eig`, does properly compute the gradients for the determinant of the matrix.

### Versions

PyTorch version: 2.0.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.35

Python version: 3.10.6 (main, May 29 2023, 11:10:38) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.42.2.el7.x86_64-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-40GB
Nvidia driver version: 470.82.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Address sizes:       43 bits physical, 48 bits virtual
Byte Order:          Little Endian
CPU(s):              128
On-line CPU(s) list: 0-127
Vendor ID:           AuthenticAMD
Model name:          AMD EPYC 7542 32-Core Processor
CPU family:          23
Model:               49
Thread(s) per core:  2
Core(s) per socket:  32
Socket(s):           2
Stepping:            0
Frequency boost:     enabled
CPU max MHz:         2900.0000
CPU min MHz:         1500.0000
BogoMIPS:            5799.92
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc art rep_good nopl nonstop_tsc extd_apicid aperfmperf eagerfpu pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_l2 cpb cat_l3 cdp_l3 hw_pstate sme retpoline_amd ssbd ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif umip overflow_recov succor smca
Virtualization:      AMD-V
L1d cache:           2 MiB (64 instances)
L1i cache:           2 MiB (64 instances)
L2 cache:            32 MiB (64 instances)
L3 cache:            256 MiB (16 instances)
NUMA node(s):        2
NUMA node0 CPU(s):   0-31,64-95
NUMA node1 CPU(s):   32-63,96-127

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==2.0.1
[pip3] torchvision==0.15.2
[pip3] triton==2.0.0
[conda] Could not collect

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @jianyuh @mruberry @walterddr @IvanYashchuk @xwang233

**Comment by x1on9f31** [comment]
I can re-produce this issue on my MacBook, the output is
```
---------torch.det----------
input matrix:
tensor([[0., 1.],
        [0., 1.]], grad_fn=<CatBackward0>)
gradients on input matrix:
tensor([[0., -0.],
        [-0., 0.]])
---------torch..eig---------
input matrix:
tensor([[0., 1.],
        [0., 1.]], grad_fn=<CatBackward0>)
gradients on input matrix:
tensor([[ 1.,  0.],
        [-1.,  0.]])
```
Below is the detailed system information
```
Collecting environment information...
PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.0 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.3 (main, Apr 19 2023, 18:49:55) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-14.0-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Max

Versions of relevant libraries:
[pip3] numpy==1.25.0
[pip3] torch==2.0.1
[pip3] torchaudio==2.0.2
[pip3] torchvision==0.15.2
[conda] numpy                     1.25.0          py311he598dae_0  
[conda] numpy-base                1.25.0          py311hfbfe69c_0  
[conda] pytorch                   2.0.1                  py3.11_0    pytorch
[conda] torchaudio                2.0.2                 py311_cpu    pytorch
[conda] torchvision               0.15.2                py311_cpu    pytorch
```

**Comment by lezcano** [comment]
The issue here is that we implement a method based on the LU decomposition as in *almost* all cases is faster and much more stable than that based on an general eigendecomposition. A general eigendecomposition is not numerically stable and is quite slow.

That being said, it's Achilles heel is matrices that are close to singular, with matrices that are exactly singular being the most annoying ones, like that from the example. I don't think there's much that we can do about that though, but ideas are welcome.

The relevant code is here:
https://github.com/pytorch/pytorch/blob/456ecefd525c3b6fa7d0854a57a3a659edc04fd1/torch/csrc/autograd/FunctionsManual.cpp#L4101-L4168

Also, this falls into the general set of edge-cases for linear algebra functions. The relevant documentation is in https://pytorch.org/docs/master/notes/numerical_accuracy.html#extremal-values-in-linalg

**Comment by lezcano** [comment]
Closing as this is expected (and documentd) due to the ill-conditioning of the computations.
